File: solutions/cpp/interest-is-interesting/1/interest_is_interesting.cpp
// interest_rate returns the interest rate for the provided balance.
double interest_rate(double balance) {

  double interest_rate{0.0};

  if (balance < 0.0) {
    interest_rate = 3.213;
  } else if ((balance >= 0) && (balance < 1000)) {
    interest_rate = 0.5;
  } else if ((balance >= 1000) && (balance < 5000)) {
    interest_rate = 1.621;

  } else {
    interest_rate = 2.475;
  }
  return interest_rate;
}

// yearly_interest calculates the yearly interest for the provided balance.
double yearly_interest(double balance) {

  return balance * ((interest_rate(balance) / 100));
}

// annual_balance_update calculates the annual balance update, taking into
// account the interest rate.
double annual_balance_update(double balance) {

  return balance + yearly_interest(balance);
}

// years_until_desired_balance calculates the minimum number of years required
// to reach the desired balance.
int years_until_desired_balance(double balance, double target_balance) {
  // guard statement for mixed signage
  if (((balance >= 0) && (target_balance < 0)) ||
      ((balance <= 0) && (target_balance > 0))) {
    return 0;
  }

  int year_counter{0};

  while (balance < target_balance) {
    balance = annual_balance_update(balance);
    year_counter++;
  }
  return year_counter;
}

File: solutions/go/interest-is-interesting/1/interest_is_interesting.go
// Package interest provides functions to calculate interest rates and balances.
package interest

// InterestRate returns the interest rate for the provided balance.
func InterestRate(balance float64) float32 {
	interestRate := float32(0)
	switch {
	case balance < 0:
		interestRate = 3.213

	case balance >= 0 && balance < 1000:
		interestRate = 0.5

	case balance >= 1000 && balance < 5000:
		interestRate = 1.621

	case balance >= 5000:
		interestRate = 2.475
	}

	return interestRate
}

// Interest calculates the interest for the provided balance.
func Interest(balance float64) float64 {
	interestGained := float64(InterestRate(balance)) * 0.01 * balance

	return interestGained
}

// AnnualBalanceUpdate calculates the annual balance update, taking into account the interest rate.
func AnnualBalanceUpdate(balance float64) float64 {
	return balance + Interest(balance)
}

// YearsBeforeDesiredBalance calculates the minimum number of years required to reach the desired balance.
func YearsBeforeDesiredBalance(balance, targetBalance float64) int {
	years := 0

	// guard statement
	if balance >= targetBalance {
		return years
	}

	for balance < targetBalance {

		balance += Interest(balance)
		years++
	}

	return years
}
